### PR TITLE
Add shared encryption controllers logic

### PR DESCRIPTION
**Pull request by enj**
Signed-off-by: Monis Khan <mkhan@redhat.com>

@sttts

### CODE REVIEW

**Comment by enj** [comment]
/retest

Unrelated flakes.

File: pkg/operator/encryption/shared_encryption.go [hunk 1]
```diff
@@ -0,0 +1,867 @@
...
+// encryptionConfSecret is the name of the final encryption config secret that is revisioned per apiserver rollout.
+// it also serves as the (map) key that is used to store the raw bytes of the final encryption config.
+const encryptionConfSecret = "encryption-config"
+
+// revisionLabel is used to find the current revision for a given API server.
+const revisionLabel = "revision"
+
+// these can only be used to encode and decode EncryptionConfiguration objects
+var (
+	encoder runtime.Encoder
```
Comment (sttts) [inline pkg/operator/encryption/shared_encryption.go]: why these? For tests?

File: pkg/operator/encryption/shared_encryption.go [hunk 2]
```diff
@@ -0,0 +1,867 @@
...
+	decoder runtime.Decoder
+)
+
+func init() {
+	scheme := runtime.NewScheme()
```
Comment (sttts) [inline pkg/operator/encryption/shared_encryption.go]: usually we make scheme and codecs global as well and only add the register calls here in init

File: pkg/operator/encryption/shared_encryption.go [hunk 3]
```diff
@@ -0,0 +1,867 @@
...
+}
+
+func (k keysState) writeKey() keyAndMode {
+	if k.writeSecret == nil {
+		return keyAndMode{}
+	}
+
+	writeKeyAndMode, _, ok := secretToKeyAndMode(k.writeSecret, k.targetNamespace)
+	if !ok {
+		klog.Infof("failed to convert write secret %s to key", k.writeSecret.Name)
```
Comment (sttts) [inline pkg/operator/encryption/shared_encryption.go]: this smells. A getter should never log, and always do what it says: return the key. If that's not always possible, add error.

File: pkg/operator/encryption/shared_encryption.go [hunk 4]
```diff
@@ -0,0 +1,867 @@
...
+
+// groupResourcesState represents the secrets (i.e. encryption keys) associated with each group resource.
+// see getDesiredEncryptionState for how this map is built.  it is first fed the current state based on the on
+// disk configuration.  the actual state of the secrets in the kube API is then layered on top to determine the
+// overall desired configuration (which is the same as the current state when the system is at steady state).
+type groupResourcesState map[schema.GroupResource]keysState
+type keysState struct {
+	targetNamespace string
+
+	readSecrets []*corev1.Secret
```
Comment (sttts) [inline pkg/operator/encryption/shared_encryption.go]: // sorted by key number, highest first

File: pkg/operator/encryption/shared_encryption.go [hunk 5]
```diff
@@ -0,0 +1,867 @@
...
+	decoder = codecs.UniversalDecoder(apiserverconfigv1.SchemeGroupVersion)
+}
+
+// groupResourcesState represents the secrets (i.e. encryption keys) associated with each group resource.
+// see getDesiredEncryptionState for how this map is built.  it is first fed the current state based on the on
+// disk configuration.  the actual state of the secrets in the kube API is then layered on top to determine the
+// overall desired configuration (which is the same as the current state when the system is at steady state).
+type groupResourcesState map[schema.GroupResource]keysState
+type keysState struct {
+	targetNamespace string
```
Comment (sttts) [inline pkg/operator/encryption/shared_encryption.go]: do we really need the targetNamespace? This looks redundant.

File: pkg/operator/encryption/shared_encryption.go [hunk 6]
```diff
@@ -0,0 +1,867 @@
...
+		if !ok {
+			klog.Infof("failed to convert read secret %s to key", readKey.Name)
+			continue
+		}
+		ret = append(ret, readKeyAndMode)
+	}
+	return ret
+}
+
+func (k keysState) writeKey() keyAndMode {
```
Comment (sttts) [inline pkg/operator/encryption/shared_encryption.go]: writeKeyAndMod

File: pkg/operator/encryption/shared_encryption.go [hunk 7]
```diff
@@ -0,0 +1,867 @@
...
+// overall desired configuration (which is the same as the current state when the system is at steady state).
+type groupResourcesState map[schema.GroupResource]keysState
+type keysState struct {
+	targetNamespace string
+
+	readSecrets []*corev1.Secret
+	writeSecret *corev1.Secret
+}
+
+func (k keysState) readKeys() []keyAndMode {
```
Comment (sttts) [inline pkg/operator/encryption/shared_encryption.go]: readKeysAndMode

File: pkg/operator/encryption/shared_encryption.go [hunk 8]
```diff
@@ -0,0 +1,867 @@
...
+	encoder = codecs.LegacyCodec(apiserverconfigv1.SchemeGroupVersion)
+	decoder = codecs.UniversalDecoder(apiserverconfigv1.SchemeGroupVersion)
+}
+
+// groupResourcesState represents the secrets (i.e. encryption keys) associated with each group resource.
+// see getDesiredEncryptionState for how this map is built.  it is first fed the current state based on the on
+// disk configuration.  the actual state of the secrets in the kube API is then layered on top to determine the
+// overall desired configuration (which is the same as the current state when the system is at steady state).
+type groupResourcesState map[schema.GroupResource]keysState
+type keysState struct {
```
Comment (sttts) [inline pkg/operator/encryption/shared_encryption.go]: keySecrets

File: pkg/operator/encryption/shared_encryption.go [hunk 9]
```diff
@@ -0,0 +1,867 @@
...
+
+func (k keysState) latestKey() (*corev1.Secret, uint64) {
+	key := k.readSecrets[0]
+	keyID, _ := secretToKeyID(key)
+	return key, keyID
+}
+
+// groupResourceKeys represents, for a single group resource, the write and read keys in a
+// format that can be directly translated to and from the on disk EncryptionConfiguration object.
+type groupResourceKeys struct {
```
Comment (sttts) [inline pkg/operator/encryption/shared_encryption.go]: this is as much about GRs as keyState is. Would align those two. `keyAndModes`?

File: pkg/operator/encryption/shared_encryption.go [hunk 10]
```diff
@@ -0,0 +1,867 @@
...
+	writeKey keyAndMode
+	readKeys []keyAndMode
+}
+
+func (k groupResourceKeys) hasWriteKey() bool {
+	return len(k.writeKey.key.Name) > 0 && len(k.writeKey.key.Secret) > 0
+}
+
+type keyAndMode struct {
```
Comment (sttts) [inline pkg/operator/encryption/shared_encryption.go]: moving all these types into `types.go` would make it all easier to read.

File: pkg/operator/encryption/shared_encryption.go [hunk 11]
```diff
@@ -0,0 +1,867 @@
...
+	for gr := range encryptionState {
+		grState := encryptionState[gr]
+		grState.readSecrets = []*corev1.Secret{writeSecret}
+		encryptionState[gr] = grState
+	}
+	klog.V(4).Infof("write secret %s set as sole write key", writeSecret.Name)
+	return encryptionState
+}
+
+type migratedGroupResources struct {
```
Comment (sttts) [inline pkg/operator/encryption/shared_encryption.go]: // migratedGroupResources is the data structured stored in the encryption.apiserver.operator.openshift.io/migrated-resources of a key secret.

**Comment by sttts** [comment]
Fixes as follow-ups.

/lgtm
/approve

**Pull request merged**